### DISCUSSION TITLE
Umi 里的 appData.bundler 在什么情况下会被更改为 mako

### DISCUSSION THREAD

**Discussion by ShotHeadman** [umijs/umi#12619] [Q&A] [answered]
我在尝试为一个修改过的umi 4框架接入mako，在构建 bundle-status.html 里面注意到构建过程里有下面一段
```tsx
 function renderStatus(data) {
        if (data.bundler === 'mako') {
          loading.setAttribute('data-mako', '');
          window.__MAKO_PERCENT = window.__MAKO_PERCENT || 0.1;
          const makoPercent = Math.floor(window.__MAKO_PERCENT * 100);
          loading.setAttribute('data-percent', makoPercent);
          window.__MAKO_PERCENT =
            window.__MAKO_PERCENT >= 1 ? 0.2 : window.__MAKO_PERCENT + 0.1;
          return;
        }
...
}
```
这里面的判断条件` if (data.bundler === 'mako')`,  data.bundler 在接入mako后似乎不会被modify，在appData这个插件初始化的时候这个值是 webpack，我也没有搜到其他地方关于这个值的改动，不过使用最新版本umi是能够正常触发mako的构建过程动画的，想问一下这个是在哪里变动的

**Comment by cklwblove** [comment, marked as answer]
1，有关 data.bundler 的值变动为 mako，可以参考代码：https://github.com/umijs/umi/blob/c81271fdf50b0f438c3359aff1d6387f29dae13a/packages/preset-umi/src/features/okam/okam.ts#L16

2，然后 bundle-status.htm 里 loadData 方法接收这里更改后的值：https://github.com/umijs/umi/blob/c81271fdf50b0f438c3359aff1d6387f29dae13a/packages/preset-umi/src/features/devTool/devTool.ts#L40
